### PR TITLE
kubeadm: use localhost for API server liveness probe

### PR DESCRIPTION
**Pull request by andrewrynhard**
**What this PR does / why we need it**:
The current liveness probe does not work with an HA cluster created with kubeadm. The probe's `host` value will be set to the IP address of the machine where `kubeadm --init` is run. For other master nodes, the IP address will be wrong.

**Special notes for your reviewer**:
None
**Release note**:
```release-note
NONE
```

/cc @timothysc

### CODE REVIEW

File: cmd/kubeadm/app/util/staticpod/utils.go [hunk 1]
```diff
@@ -184,9 +184,7 @@ func WriteStaticPodToDisk(componentName, manifestDir string, pod v1.Pod) error {
 func GetProbeAddress(cfg *kubeadmapi.MasterConfiguration, componentName string) string {
 	switch {
 	case componentName == kubeadmconstants.KubeAPIServer:
-		if cfg.API.AdvertiseAddress != "" {
-			return cfg.API.AdvertiseAddress
-		}
+		return "127.0.0.1"
```
Comment (andrewrynhard) [inline cmd/kubeadm/app/util/staticpod/utils.go]: I don't believe we can use DownwardAPI here. Correct me if I am wrong. If we could, I would rather use the `$(HOST_IP)` implemented in https://github.com/kubernetes/kubernetes/commit/49e01b05e7791d8bbb4863f83c0b2f420cec73ab.

Comment (timothysc) [reply cmd/kubeadm/app/util/staticpod/utils.go]: Maybe I'm missing something but I don't get the argument you made for the initial PR re: HA. 

/cc @jamiehannaford

Comment (andrewrynhard) [reply cmd/kubeadm/app/util/staticpod/utils.go]: I created a cluster using https://kubernetes.io/docs/setup/independent/high-availability. I also enabled the `SelfHosting` feature. The `DaemonSet` the defines the API server is setup on a "seed" host, which sets up the liveness probes in the pod spec to use `cfg.API.AdvertiseAddress`.  The `cfg.API.AdvertiseAddress` value is automatically set to the seed host's IP. So for masters that are not the seed host, you end up with liveness probes using host that is not their own. Hope that clears it up.

**Comment by jamiehannaford** [comment]
The issue I posted here https://github.com/kubernetes/kubeadm/issues/583 was only occuring when you specified the LB IPv4 as the advertise address. The health checks were failing because the LB was balancing to masters without a functioning apiserver (therefore failing). We changed the HA guide to use the masters' private IPv4s as the advertise IPs instead, so I understand why this is an issue.

I'm not too sure what you mean by "seed host" @andrewrynhard, can you expand? My understanding was that each master that's added listens on its own private IPv4, which is accessible by health checks.

**Comment by andrewrynhard** [comment]
@jamiehannaford

The "seed host" is the host where `kubeadm init` is run when using the `SelfHosting` feature. The HA [guide](https://kubernetes.io/docs/setup/independent/high-availability) assumes that users will not be using the `SelfHosting` feature, and thus informs the user to run `kubeadm init` on each machine destined to be a master. In the guide the `MasterConfiguration` is modified on each of the hosts to reflect that host's IP in the `advertiseAddress` field. 

In the case that `SelfHosting` is enabled, you run `kubeadm init` only on one host, then other hosts that come up with the `node-role.kubernetes.io/master` label will become masters. The issue is that the `DaemonSet` doing this is created on the seed host, and therefore masters that are not the seed host will have the `advertiseAddress` field of the seed host.

I hope that clarifies the issue :)

**Comment by jamiehannaford** [comment]
@timothysc wdyt? ^

**Comment by andrewrynhard** [comment]
@dixudx Looks like I have a test to fix.

**Comment by andrewrynhard** [comment]
@jamiehannaford I have updated this PR such that 127.0.0.1 will be used only when `SelfHosting` is `true`.

File: cmd/kubeadm/app/util/staticpod/utils.go [hunk 2]
```diff
@@ -184,7 +185,9 @@ func WriteStaticPodToDisk(componentName, manifestDir string, pod v1.Pod) error {
 func GetProbeAddress(cfg *kubeadmapi.MasterConfiguration, componentName string) string {
 	switch {
 	case componentName == kubeadmconstants.KubeAPIServer:
-		if cfg.API.AdvertiseAddress != "" {
+		if features.Enabled(cfg.FeatureGates, features.SelfHosting) {
```
Comment (timothysc) [inline cmd/kubeadm/app/util/staticpod/utils.go]: Please add a comment block here with the details you outlined, it is pretty unclear on 1st inspection why

**Review by timothysc — CHANGES_REQUESTED** [review]
/approve 

Please address comment, then lgtm

File: cmd/kubeadm/app/util/staticpod/utils.go (hunk 2 above)
Comment (andrewrynhard) [reply cmd/kubeadm/app/util/staticpod/utils.go]: 👍

**Comment by dixudx** [comment]
@andrewrynhard  Please squash your commits. No need to have two commits for such a small change.

**Review by timothysc — APPROVED** [review]
/lgtm

**Review by timothysc — APPROVED** [review]
/lgtm

**Pull request merged**